// action_bar/index.js
const { terminal, wand, help } = require('icons')
const bar = require('action_bar/search_bar')

module.exports = action_bar

function action_bar () {
  const search_bar = bar('Home > Documents > Current > theme')

  // creating parent
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  // styling
  const sheet = new CSSStyleSheet()
  const opts = {}
  sheet.replaceSync(get_theme(opts))
  shadow.adoptedStyleSheets = [sheet]

  shadow.innerHTML = `
  <div class="action-bar-container">
    <div class="action-bar-content">
      <button class="icon-button">
        ${terminal()}
      </button>
      <div class="separator"></div>
      <button class="icon-button">
        ${wand()}
      </button>
      <div class="separator"></div>
      <searchbar></searchbar>
      <button class="icon-button">
        ${help()}
      </button>
    </div>
  </div>`

  shadow.querySelector('searchbar').replaceWith(search_bar)

  // to add a click event listener to the buttons:
  // const [btn1, btn2, btn3] = shadow.querySelectorAll('button')
  // btn1.addEventListener('click', () => { console.log('Terminal button clicked') })

  return el
}

function get_theme () { // opts
  return `
  .action-bar-container {
      display: flex;
      align-items: center;
      background-color: #212121;
      padding: 0.5rem;
      // min-width: 456px
  }

  .action-bar-content {
      display: flex;
      align-items: center;
      gap: 0.5rem;
      flex:1;
  }

  .icon-button {
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 0;
    border: none;
    background-color: transparent;
    cursor: pointer;
  }


  .separator {
      width: 1px;
      height: 24px;
      background-color: #424242;
  }

  .search-bar-container {
    flex: 1;
    position: relative;
  }
  svg {
    display: block;
    margin: auto;
  }
`
}
