// Action-Bar/index.js
const { terminal, wand, help } = require('icons')
const SearchBar = require('Action-Bar/search-bar')

module.exports = action_bar

function action_bar () {
  const search_bar = SearchBar('Home > Documents > Current > theme')

  // creating parent
  const el = document.createElement('div')
  el.className = 'action-bar-container'
  el.id = 'action_bar'
  const shadow = el.attachShadow({ mode: 'closed' })

  // styling
  const sheet = new CSSStyleSheet()
  const opts = {}
  sheet.replaceSync(get_theme(opts))
  shadow.adoptedStyleSheets = [sheet]

  shadow.innerHTML = `
  <div class = 'action-bar-container'>
    <div class="action-bar-content">
        <button class='icon-button' id="Open Terminal">
            ${terminal()}
        </button>
        <div class='separator'></div>
        <button class='icon-button' id="Magic Wand">
            ${wand()}
        </button>
        <div class='separator'></div>
        <button class='icon-button' id="Help">
            ${help()}
        </button>
    </div>
  </div>
  `

  const separator_element = shadow.querySelector('.separator:last-of-type')
  shadow.querySelector('.action-bar-content').insertBefore(search_bar, separator_element)

  // to add a click event listener to the buttons:
  // const terminal_button = shadow.querySelector(`.icon-button[id="Open Terminal"]`)
  // terminal_button.addEventListener('click', () => { console.log('Terminal button clicked') })

  return el
}

function get_theme () { // opts
  return `
  .action-bar-container {
      display: flex;
      align-items: center;
      background-color: #212121;
      padding: 0.5rem;
  }

  .action-bar-content {
      display: flex;
      align-items: center;
      gap: 0.5rem;
      flex:1;
  }

  .icon-button {
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 0;
    border: none;
    background-color: transparent;
    cursor: pointer;
  }


  .separator {
      width: 1px;
      height: 24px;
      background-color: #424242;
  }

  .search-bar-container {
    flex: 1;
    position: relative;
  }
  svg {
    display: block;
    margin: auto;
  }
`
}
