// Action-Bar/index.js
const ActionBar = require('./action-bar');

function Page() {
  const pageContainerClass = 'pageContainer';
  const pageContentClass = 'pageContent';
  const pageHeadingClass = 'pageHeading';
  const pageParagraphClass = 'pageParagraph';

  const container = document.createElement('div');
  container.className = pageContainerClass;

  const actionBar = ActionBar();
  container.appendChild(actionBar);

  const contentDiv = document.createElement('div');
  contentDiv.className = pageContentClass;

  const heading = document.createElement('h1');
  heading.className = pageHeadingClass;
  heading.textContent = 'Content Area';
  contentDiv.appendChild(heading);

  const paragraph = document.createElement('p');
  paragraph.className = pageParagraphClass;
  paragraph.textContent = 'Click on the search bar above to see it in action!';
  contentDiv.appendChild(paragraph);
  container.appendChild(contentDiv);
  const styleElement = document.createElement('style');
  styleElement.textContent = style;
  container.appendChild(styleElement); 

  return container;

}

module.exports = Page;
const style = `
:root {
  background-color: #000000
}

.actionBarContainer {
  height: 3rem;
  padding-left: 1rem;
  padding-right: 1rem;
  margin: 0;
  display: flex;
  align-items: center;
  background-color: #1a1a1a;
  gap: 0.5rem;
  border: 0.068rem solid #464646;
}

.iconButton {
  background-color: #1a1a1a;
  transition: background-color 0.3s;
  border: 0;
  shadow: 0;
  display: flex;
  align-items: center;
  flex-direction: column;
}

.iconButton:hover {
  cursor: pointer;
  }
.iconButton:active {
  
}
.separator {
  width: 1px;
  height: 1.5rem;
  background-color: #e5e7eb;
  margin-left: 0.5rem;
  margin-right: 0.5rem;
}

.searchBarContainer {
  flex: 1;
  position: relative;
}

.searchInputContainer {
  height: 2rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  display: flex;
  align-items: center;
  justify-content: space-between;
  background-color: #303030;
  border-radius: 0.375rem;
  cursor: text;
}

.searchInputContent {
  flex: 1;
}

.searchInputText {
  font-size: 0.875rem;
  color: #a0a0a0;
}

.searchInput {
  width: 100%;
  background-color: transparent;
  outline: none;
  border: none;
  color: #a0a0a0;
  font-size: 0.875rem;
}

.searchResetButton {
  margin-left: 0;
  padding: 0;
  border: none;
  background-color: transparent;
}

.searchResetButton:hover {
  background-color: #e5e7eb;
}

.pageContainer {
  display: flex;
  flex-direction: column-reverse;
  height: 100vh;
  justify-content: space-between;
  min-height: 100vh;
  background-color: black;
}

.pageContent {
  padding: 1rem;
}

.pageHeading {
  font-size: 1.5rem;
  font-weight: bold;
}

.pageParagraph {
  margin-top: 0.5rem;
  color: #6b7280;
}
`
